I built from the branch in this pull request (rebased on blead) in 4 different ways and got slight, perhaps significant results in the output.  (As I repeated the process, my commands got a bit more precisely focused.)

* GH 23661 "regular" build using gcc-13 on Ubuntu Linux 24.04 LTS ("regular" means simply `-des -Dusedevel`.)
```
$ cd t;./perl harness -v run/todo.t | ack '13307'; cd -
not ok 8 - Hashref element reference in @_ disappeared; GH 13307 # TODO GH 13307
```

* GH 23661 "regular" build using clang-18 on Ubuntu Linux 24.04 LTS
```
$ cd t;./perl harness -v run/todo.t | ack '(^Result:|13307)'; cd -
not ok 8 - Hashref element reference in @_ disappeared; GH 13307 # TODO GH 13307
# Failed test 8 - Hashref element reference in @_ disappeared; GH 13307 at run/todo.t line 228
Result: PASS
```
* GH 23661 "-DPERL_RC_STACK" build using gcc-13 on Ubuntu Linux 24.04 LTS
```
$ sh ./Configure -des -Dusedevel -Dcc=gcc -Accflags=-DPERL_RC_STACK && make test_prep
...
$ cd t;./perl harness -v run/todo.t | ack '(^Result:|13307)'; cd -
ok 8 - Hashref element reference in @_ disappeared; GH 13307
Result: PASS
```
* GH 23661 "-DPERL_RC_STACK" build using clang-18 on Ubuntu Linux 24.04 LTS
```
$ sh ./Configure -des -Dusedevel -Dcc=clang -Accflags=-DPERL_RC_STACK 1>/dev/null && make test_prep 1>/dev/null
pp_hot.c:3147:21: warning: variable 'i' set but not used [-Wunused-but-set-variable]
 3147 |             SSize_t i;
      |                     ^
1 warning generated.
Warning (mostly harmless): No library found for -lposix
Warning (mostly harmless): No library found for -lcposix
...
$ ok 8 - Hashref element reference in @_ disappeared; GH 13307
Result: PASS
```
Note the build-time warning in the 4th invocation.  I spent several hours puzzling over this last night.  AFAICT it is only emitted with a (recent?) `clang` build and only with `-Accflags=-DPERL_RC_STACK`.  I tried to eliminate the warning with this combination of compiler and config options, but was unsuccessful, probably because I don't have much experience with the implications of `#ifdef`, `#ifndef` and `#endif` are for scoping of variables in C.

Note further that the way @tonycoz has suggested checking for `-DPERL_RC_STACK` differs from the only other instance of checking for that config option I could find underneath `t/`, that being in `t/refcount.t`:
```
 22 BEGIN {
 23     chdir 't' if -d 't';
 24     require './test.pl';
 25     skip_all('not built with PERL_RC_STACK')
 26         unless defined &Internals::stack_refcounted
 27             && (Internals::stack_refcounted() & 1);
 28     set_up_inc( qw(. ../lib) );
 29 }
```

--- a/.mailmap
+++ b/.mailmap
@@ -504,6 +504,7 @@ Juerd Waalboer <#####@juerd.nl> Juerd <juerd@convolution.nl>
 Juerd Waalboer <#####@juerd.nl> Juerd <juerd@cpan.org>
 Juerd Waalboer <#####@juerd.nl> Juerd Waalboer <juerd@c3.convolution.nl>
 Justin Case <unknown> Justin Case <>
+Jörg Thomas <zebster@spr.at> zb226 <zebster@spr.at>
 Kaoru Maeda <maeda@src.ricoh.co.jp> Unknown Ricoh Contributor II <maeda@src.ricoh.co.jp>
 Karen Etheridge <ether@cpan.org> Karen Etheridge <github@froods.org>
 Karl Williamson <khw@cpan.org> Karl <khw@karl.(none)>

--- a/AUTHORS
+++ b/AUTHORS
@@ -780,6 +780,7 @@ juna                           <ggl.20.jj...@spamgourmet.com>
 Jungshik Shin                  <jshin@mailaps.org>
 Justin Banks                   <justinb@cray.com>
 Justin Case
+Jörg Thomas                    <zebster@spr.at>
 Jörg Walter                    <jwalt@cpan.org>
 Ka-Ping Yee                    <kpyee@aw.sgi.com>
 kafka                          <kafka@madrognon.net>

--- a/t/run/todo.t
+++ b/t/run/todo.t
@@ -206,6 +206,29 @@ TODO: {
 }
 
 TODO: {
+    my $is_rc_stack = "$Config{cc} $Config{ccflags} $Config{optimize}" =~ /-DPERL_RC_STACK\b/;
+    local $::TODO = $is_rc_stack ? undef : "GH 13307";
+
+    my $results = fresh_perl(<<~'EOF', {});
+        my $iter;
+        my %llll;
+        sub bbbb { }
+        sub aaaa {
+            \@_ if $iter & 1;
+            bbbb($_[1]) if $iter & 2;
+            delete $llll{p};
+            print $_[0] // "undef", "\n";
+        }
+        for(0..3) {
+            $iter = $_;
+            %llll = (p => "oooo");
+            aaaa($llll{p}, undef);
+        }
+    EOF
+    is($results, "oooo\noooo\noooo\noooo", 'Hashref element reference in @_ disappeared; GH 13307');
+}
+
+TODO: {
     local $::TODO = "GH 16008";
     my $results = fresh_perl(<<~'EOF', {} );
         open my $h, ">", \my $x;

